src/rdf4cpp/util/Int128.hpp

```diff
@@ -0,0 +1,425 @@
+            return std::to_chars_result{.ptr = first, .ec = std::errc::value_too_large};
+        }
+
+        *first++ = '-';
+        return to_chars(first, last, static_cast<unsigned __int128>(-val));
+    }
+    inline bool to_chars_canonical(__int128 const value, writer::BufWriterParts const writer) noexcept {
+        // +1 because of definition of digits10 https://en.cppreference.com/w/cpp/types/numeric_limits/digits10
+        // +1 for sign
+        static constexpr size_t buf_sz = std::numeric_limits<__int128>::digits10 + 1 + 1;
+
+        std::array<char, buf_sz> buf;
+        std::to_chars_result const res = to_chars(buf.data(), buf.data() + buf.size(), value);
+
+        assert(res.ec == std::errc{});
+
+        std::string_view const s{buf.data(), static_cast<std::string::size_type>(res.ptr - buf.data())};
+        return writer::write_str(s, writer);
+    }
+#endif
+
+    namespace detail {
+        enum struct OverflowMode {
+            Checked,
+            UndefinedBehavior,
+        };
+
+        template<std::size_t MinBits, std::size_t MaxBits, boost::multiprecision::cpp_integer_type SignType, typename Alloc>
+        using cpp_int_checked = boost::multiprecision::number<boost::multiprecision::cpp_int_backend<MinBits, MaxBits, SignType, boost::multiprecision::checked, Alloc>>;
+        template<std::size_t MinBits, std::size_t MaxBits, boost::multiprecision::cpp_integer_type SignType, typename Alloc>
+        using cpp_int_unchecked = boost::multiprecision::number<boost::multiprecision::cpp_int_backend<MinBits, MaxBits, SignType, boost::multiprecision::unchecked, Alloc>>;
+
+        template<typename T>
+        concept IntegralExt = std::integral<T> || std::same_as<T, __int128> || std::same_as<T, unsigned __int128>;
+        template<typename T>
+        concept BoostNumber = std::same_as<T, boost::multiprecision::number<typename T::backend_type, T::et>>;
+
+        template<OverflowMode m, typename T>
+        requires IntegralExt<T>
+        static constexpr bool add_checked(T const &a, T const &b, T &result) noexcept {
+            if constexpr (m == OverflowMode::Checked) {
+                return __builtin_add_overflow(a, b, &result);
+            } else {
+                result = a + b;
+                return false;
+            }
+        }
+        template<OverflowMode m, std::size_t MinBits, std::size_t MaxBits, boost::multiprecision::cpp_integer_type SignType, typename Alloc>
+        static constexpr bool add_checked(cpp_int_checked<MinBits, MaxBits, SignType, Alloc> const &a,
+                                          cpp_int_checked<MinBits, MaxBits, SignType, Alloc> const &b,
+                                          cpp_int_checked<MinBits, MaxBits, SignType, Alloc> &result) noexcept {
+            if (m == OverflowMode::Checked) {
+                try {
+                    result = a + b;
+                } catch (std::overflow_error const &) {
+                    return true;
```

Comment:
these return values are kind of the wrong way around for the function name. That should at least be document (maybe with a comment at the top of the file)